File: Client/file-exploder/Utilities/FormatUtils.swift
import Foundation

struct FormatUtils {
    /// Reused across every row: building a `DateFormatter` per call is costly
    /// enough to show up while scrolling a large directory.
    private static let dateFormatter: DateFormatter = {
        let formatter = DateFormatter()
        formatter.dateStyle = .medium
        formatter.timeStyle = .short
        return formatter
    }()

    /// Reused across every row for the same reason as dateFormatter above.
    private static let sizeFormatter: ByteCountFormatter = {
        let formatter = ByteCountFormatter()
        formatter.countStyle = .file
        return formatter
    }()

    static func formattedSize(_ bytes: Int64) -> String {
        sizeFormatter.string(fromByteCount: bytes)
    }

    static func formattedDate(_ date: Date) -> String {
        dateFormatter.string(from: date)
    }
}

struct FileIcons {
    static func icon(for file: RemoteFile) -> String {
        if file.isDirectory {
            return "folder.fill"
        }
        if file.isSymlink {
            return "link"
        }

        let ext = (file.name as NSString).pathExtension.lowercased()
        
        switch ext {
        // Images
        case "jpg", "jpeg", "png", "gif", "svg", "webp", "bmp", "tiff":
            return "photo"
        // Documents
        case "pdf":
            return "doc.richtext"
        case "doc", "docx":
            return "doc"
        case "xls", "xlsx":
            return "tablecells"
        case "ppt", "pptx":
            return "rectangle.on.rectangle"
        // Archives
        case "zip", "tar", "gz", "7z", "rar", "bz2", "xz":
            return "archivebox"
        // Audio
        case "mp3", "wav", "aac", "flac", "ogg", "m4a":
            return "music.note"
        // Video
        case "mp4", "mov", "avi", "mkv", "webm", "flv":
            return "film"
        // Text
        case "txt", "md", "log", "rtf":
            return "doc.text"
        // Code
        case "json", "xml", "plist", "yaml", "yml", "toml":
            return "doc.plaintext"
        case "sh", "bash", "zsh", "fish":
            return "terminal"
        case "py", "rb", "js", "ts", "swift", "go", "rs", "java", "c", "cpp", "h":
            return "chevron.left.forwardslash.chevron.right"
        // Executables
        case "exe", "app", "dmg", "deb", "rpm":
            return "app"
        // Links
        case "link", "url", "webloc":
            return "link"
        default:
            return "doc"
        }
    }
}

extension String {
    /// Escapes a string to be safely used in a Unix shell command.
    /// Wraps the string in single quotes and escapes internal single quotes.
    var shellEscaped: String {
        let escaped = self.replacingOccurrences(of: "'", with: "'\\''")
        return "'\(escaped)'"
    }

    /// ASCII-only transport for remote paths. This prevents macOS process
    /// argument conversion from changing composed characters into an
    /// equivalent normalization that names a different file on Linux.
    var utf8Base64: String {
        Data(utf8).base64EncodedString()
    }
}

File: WindowsClient/FileExploder/Utilities/FormatUtils.cs
using FileExploder.Models;

namespace FileExploder.Utilities;

public static class FormatUtils
{
    public static string FormattedSize(long bytes)
    {
        const long kb = 1_000;
        const long mb = kb * 1_000;
        const long gb = mb * 1_000;
        const long tb = gb * 1_000;

        if (bytes < kb)
        {
            return $"{bytes} bytes";
        }

        var (value, unit) = bytes switch
        {
            >= tb => (bytes / (double)tb, "TB"),
            >= gb => (bytes / (double)gb, "GB"),
            >= mb => (bytes / (double)mb, "MB"),
            _ => (bytes / (double)kb, "KB"),
        };

        // Roughly three significant digits, the way most file managers show
        // sizes: "1.2 GB", "45.3 MB", "999 KB".
        var decimals = value >= 100 ? 0 : value >= 10 ? 1 : 2;
        return $"{value.ToString("F" + decimals)} {unit}";
    }

    public static string FormattedDate(DateTimeOffset date) =>
        date.ToLocalTime().ToString("g");
}

public static class FileIcons
{
    /// A short glyph standing in for a proper icon set: this client has no
    /// bundled vector icon assets, so a Unicode symbol carries the same
    /// at-a-glance file-type information a real icon would.
    public static string Icon(RemoteFile file)
    {
        if (file.IsDirectory)
        {
            return "📁";
        }
        if (file.IsSymlink)
        {
            return "🔗";
        }

        var ext = System.IO.Path.GetExtension(file.Name).TrimStart('.').ToLowerInvariant();
        return ext switch
        {
            "jpg" or "jpeg" or "png" or "gif" or "svg" or "webp" or "bmp" or "tiff" => "🖼️",
            "pdf" => "📕",
            "doc" or "docx" => "📄",
            "xls" or "xlsx" => "📊",
            "ppt" or "pptx" => "📽️",
            "zip" or "tar" or "gz" or "7z" or "rar" or "bz2" or "xz" => "🗜️",
            "mp3" or "wav" or "aac" or "flac" or "ogg" or "m4a" => "🎵",
            "mp4" or "mov" or "avi" or "mkv" or "webm" or "flv" => "🎬",
            "txt" or "md" or "log" or "rtf" => "📝",
            "json" or "xml" or "plist" or "yaml" or "yml" or "toml" => "🗒️",
            "sh" or "bash" or "zsh" or "fish" => "💻",
            "py" or "rb" or "js" or "ts" or "swift" or "go" or "rs" or "java" or "c" or "cpp" or "h" => "🧩",
            "exe" or "app" or "dmg" or "deb" or "rpm" => "⚙️",
            "link" or "url" or "webloc" => "🔗",
            _ => "📄",
        };
    }
}
